scipy/optimize/_highs/pyHiGHS/src/highs_wrapper.pyx

```diff
@@ -0,0 +1,622 @@
+cdef HighsOptions _ref_opts
+cdef cppmap[string, OptionRecord*] _ref_opt_lookup
+cdef OptionRecord * _r = NULL
+for _r in _ref_opts.records:
+    _ref_opt_lookup[_r.name] = _r
+
+cdef str _opt_warning(string name, val, valid_set=None):
+    cdef OptionRecord * r = _ref_opt_lookup[name]
+
+    # BOOL
+    if r.type == HighsOptionTypeBOOL:
+        default_value = (<OptionRecordBool*> r).default_value
+        return ('Option "%s" is "%s", but only True or False is allowed. '
+                'Using default: %s.' % (name.decode(), str(val), default_value))
+
+    # INT
+    if r.type == HighsOptionTypeINT:
+        lower_bound = int((<OptionRecordInt*> r).lower_bound)
+        upper_bound = int((<OptionRecordInt*> r).upper_bound)
+        default_value = int((<OptionRecordInt*> r).default_value)
+        if upper_bound - lower_bound < 10:
+            int_range = str(set(range(lower_bound, upper_bound + 1)))
+        else:
+            int_range = '[%d, %d]' % (lower_bound, upper_bound)
+        return ('Option "%s" is "%s", but only values in %s are allowed. '
+                'Using default: %d.' % (name.decode(), str(val), int_range, default_value))
+
+    # DOUBLE
+    if r.type == HighsOptionTypeDOUBLE:
+        lower_bound = (<OptionRecordDouble*> r).lower_bound
+        upper_bound = (<OptionRecordDouble*> r).upper_bound
+        default_value = (<OptionRecordDouble*> r).default_value
+        return ('Option "%s" is "%s", but only values in (%g, %g) are allowed. '
+                'Using default: %g.' % (name.decode(), str(val), lower_bound, upper_bound, default_value))
+
+    # STRING
+    if r.type == HighsOptionTypeSTRING:
+        if valid_set is not None:
+            descr = 'but only values in %s are allowed. ' % str(set(valid_set))
+        else:
+            descr = 'but this is an invalid value. %s. ' % r.description.decode()
+        default_value = (<OptionRecordString*> r).default_value.decode()
+        return ('Option "%s" is "%s", '
+                '%s'
+                'Using default: %s.' % (name.decode(), str(val), descr, default_value))
+
+    # We don't know what type (should be unreachable)?
+    return('Option "%s" is "%s", but this is not a valid value. '
+           'See documentation for valid options. '
+           'Using default.' % (name.decode(), str(val)))
+
+cdef apply_options(dict options, Highs & highs):
+    '''Take options from dictionary and apply to HiGHS object.'''
+
+    # Send logging to dummy file to get rid of output from stdout
+    if options.get('message_level', None) == <int> ML_NONE:
+        highs.setHighsLogfile(NULL)
+        highs.setHighsOutput(NULL)
+    else:
+        # Empty file to send to stdout
+        highs.setHighsLogfile(stdout)
+        highs.setHighsOutput(stdout)
+
+    # Initialize for error checking
+    cdef HighsStatus opt_status = HighsStatusOK
+
+    # Do all the ints
+    for opt in set([
+            b'allowed_simplex_cost_scale_factor',
```

Comment:
What's the deal with creating these as bytes? It looks odd, so would deserve at least a comment. And I would probably prefer encoding at the point where it's needed right when it goes into the C++ code.

Reply:
Changed to prefer `encode()`